### PR TITLE
Enumerable Extensions - Single and SingleOrDefault

### PR DESCRIPTION
**Pull request by ckimball**
adding enumerable extension where you really want to perform a single, but safely

### CODE REVIEW

File: FunctionalProgramming/Basics/EnumerableExtensions.cs [hunk 1]
```diff
@@ -231,6 +230,31 @@ public static IMaybe<T> MaybeFirst<T>(this IQueryable<T> ts, Expression<Func<T,
...
+        /// Type-safe version of LINQs 'Single' and 'SingleOrDefault' function that accepts a predicate, returning
+        /// a Maybe instead of possibly throwing an exception or returning null
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="ts">An enumerable from which we want the only element that satisfies the provided predicate</param>
+        /// <param name="predicate">A predicate for which only one element will satisfy to be the return value</param>
+        /// <returns>The only value in the enumerable that satisfies the predicate, or nothing if no such element exists or multiple elements satisifying the predicate exist</returns>
+        public static IMaybe<T> MaybeSingle<T>(this IEnumerable<T> ts, Func<T, bool> predicate) where T : class
+        {
+            return Try.Attempt(() => ts.SingleOrDefault(predicate).ToMaybe()).GetOrElse(Maybe.Nothing<T>);
```
Comment (bkyrlach) [inline FunctionalProgramming/Basics/EnumerableExtensions.cs]: Consider the following...

```return Try.Attemp(() => ts.SingleOrDefault(predicate).ToMaybe()).AsMaybe().Join();```

**Comment by bkyrlach** [comment]
You may also want to include an implementation for ```IQueryable```, otherwise you might find the performance undesirable when using unrealized queries.

File: FunctionalProgramming/Basics/EnumerableExtensions.cs (hunk 1 above)
Comment (ckimball) [inline FunctionalProgramming/Basics/EnumerableExtensions.cs]: It says that .Join() doesn't exist.

**Comment by ckimball** [comment]
How about that?

**Pull request merged**